index.js

```diff
@@ -30,6 +30,22 @@ function attachComments(text, ast, opts) {
   return astComments;
 }
 
+// Cache compiled regexps in the likely case the same pragma is used for many
+// files.
+const reForPragma = {};
+function hasPragma(astComments, pragma) {
+  let re;
+  if (reForPragma[pragma] != null) {
+    re = reForPragma[pragma];
+  } else {
+    // @pragma surrounded by any amount of whitespace on either or both sides
+    re = reForPragma[pragma] = new RegExp(`\\s*@${pragma}\\s*`);
+  }
+
+  const firstComment = astComments[0];
```

Comment:
Why does it have to be the first comment? There are lots of magic comments you might start your files with:

- `@prettier`
- `@flow`
- `eslint-disable no-console`
- licensing stuff

Reply:
All the fb magic comments only work if it's the first comment of the file.

Reply:
it doesn't need to be the first thing in the comment, but it needs to be a part of the first comment, like in a docblock. You an still have other pragmas and licensing information within that first docblock (see most open source FB projects)

Reply:
There are two practical reasons for this:
1) We don't need to parse the file, we can just grep until we see */ to find it which helps making it fast.
2) We use them to toggle syntax variants, so we can't actually parse the file until we know which variant to use.

We're not doing any of those two right now with prettier but those are two optimizations that would be useful.

Reply:
Good, sounds like this has been thought through then. 👍

Reply:
Actually, we should probably implement 1), otherwise it's going to throw error for js files that prettier doesn't understand or be super slow for giant generated files where it doesn't need to.

Reply:
I played around a little bit with Flow, and this works:

```js
/*
function notCode() {
  return 1;
}
*/
// @prettier

// something something @flow side

code;
```

It appears as if `@flow` can be present anywhere in any comment that comes before all code. Additionally, `@flow` seems to have to be located at the very beginning of a comment or line, at the very end of a line, or surrounded by whitespace.